### PR TITLE
[CBT] Implement VDI.get_nbd_info XenAPI call

### PR DESCRIPTION
**Pull request by gaborigloi**

### CODE REVIEW

File: ocaml/xapi/test_common.ml [hunk 1]
```diff
@@ -359,3 +359,7 @@ let make_pool_update ~__context ?(ref=Ref.make ()) ?(uuid=make_uuid ())
   let update_info = Xapi_pool_update.{uuid; name_label; name_description; version; key; installation_size; after_apply_guidance} in
   Xapi_pool_update.create_update_record ~__context ~update:ref ~update_info ~vdi;
   ref
+
+let make_session ~__context ?(ref=Ref.make ()) ?(uuid=make_uuid ()) ?(this_host=Ref.null) ?(this_user=Ref.null) ?(last_active=API.Date.never) ?(pool=false) ?(other_config=[]) ?(is_local_superuser=false) ?(subject=Ref.null) ?(validation_time=API.Date.never) ?(auth_user_sid="") ?(auth_user_name="") ?(rbac_permissions=[]) ?(parent=Ref.null) ?(originator="test") () =
```
Comment (thomassa) [inline ocaml/xapi/test_common.ml]: Is this function copied from (or based on) an existing one? If so, why do we need this new one?

Comment (gaborigloi) [inline ocaml/xapi/test_common.ml]: I took the inspiration from `Test_event.create_session` - shall I update that code to use this helper function?

Comment (thomassa) [inline ocaml/xapi/test_common.ml]: That would make sense.

File: ocaml/xapi/test_vdi_cbt.ml [hunk 2]
```diff
@@ -62,9 +62,95 @@ let test_set_metadata_of_pool_doesnt_allow_cbt_metadata_vdi () =
...
+    host
+  in
+
+  let setup_test () =
+    let __context = Mock.make_context_with_new_db "Mock context" in
+    let session_id = Test_common.make_session ~__context () in
+    let __context = Context.make ~__context ~session_id "test" in
+    let make_host = make_host __context in
+
+    let sr_of_vdi = Test_common.make_sr ~__context () in
```
Comment (thomassa) [inline ocaml/xapi/test_vdi_cbt.ml]: Slightly confusing name: `sr_of_vdi` looks as if it should be a function.
`sr_for_vdi` would be clearer, at least for me.

Comment (gaborigloi) [inline ocaml/xapi/test_vdi_cbt.ml]: I wanted to say that this is the SR that stores the VDI.

File: ocaml/xapi/xapi_vdi.ml [hunk 3]
```diff
@@ -861,6 +861,35 @@ let export_changed_blocks ~__context ~vdi_from ~vdi_to =
...
+    )
+
+let get_nbd_info ~__context ~self =
+  if (Db.VDI.get_type ~__context ~self) = `cbt_metadata then begin
+    error "VDI.get_nbd_info: called with a VDI of type cbt_metadata (at %s)" __LOC__;
+    raise (Api_errors.Server_error (Api_errors.vdi_incompatible_type, [ Ref.string_of self; Record_util.vdi_type_to_string `cbt_metadata ]))
+  end;
+
+  let sr = Db.VDI.get_SR ~__context ~self in
+  let hosts_with_attached_pbds =
```
Comment (thomassa) [inline ocaml/xapi/xapi_vdi.ml]: The code to populate `hosts_with_attached_pbds` looks correct.
Using one of the `get_..._where` functions might be faster but this is not going to be called often so it doesn't matter.

File: ocaml/xapi/xapi_vdi.ml [hunk 4]
```diff
@@ -861,6 +861,35 @@ let export_changed_blocks ~__context ~vdi_from ~vdi_to =
...
+    Db.SR.get_PBDs ~__context ~self:sr
+    |> List.filter (fun pbd -> Db.PBD.get_currently_attached ~__context ~self:pbd)
+    |> List.map (fun pbd -> Db.PBD.get_host ~__context ~self:pbd)
+  in
+  let get_ips host =
+    let get_ips pif =
+      (Db.PIF.get_IP ~__context ~self:pif) :: (Db.PIF.get_IPv6 ~__context ~self:pif)
```
Comment (thomassa) [inline ocaml/xapi/xapi_vdi.ml]: Consider omitting the first item from this list if `get_IP` returns the empty string.
(Equivalently, pipe the whole lot of `ips` through a filter to remove any empty strings.)

**Comment by thomassa** [comment]
Looks good!
I think we should also have at least one test-case in which there will be multiple URIs for the same host (a host with multiple PIFs).
Also hosts with IPv6 addresses.

Perhaps also a host with multiple PBDs attaching it to the same SR (if that's possible)? Check that this doesn't give rise to multiple identical URIs.

**Comment by gaborigloi** [comment]
@thomassa I have realised that I should maybe add more try/with blocks to this code to check for null errors. According to https://xapi-project.github.io/xen-api/ and the source code of xapi, it is possible to create a `PIF` with a `NULL` `host` using `PIF.db_introduce`, and maybe even `PIF.introduce`. It also seems to be possible to create a `PBD` with a `NULL` `host` using `PBD.create`, because it doesn't check whether `host` is `NULL`.

**Comment by gaborigloi** [comment]
@thomassa it seems that the code doesn't allow one to have multiple PBDs attaching the same SR to the same host: https://github.com/xapi-project/xen-api/blob/master/ocaml/xapi/xapi_pbd.ml#L33

File: ocaml/xapi/xapi_vdi.ml (hunk 4 above)
Comment (gaborigloi) [inline ocaml/xapi/xapi_vdi.ml]: The IPs can be temporarily empty if they are configured via DHCP. And PIFs may not even have an IP, in which case it is the empty string again.

**Comment by gaborigloi** [comment]
Closed as it needs some work to make it more robust.

**Comment by thomassa** [comment]
I'm happy for this to be merged once the automatic check passes.

**Pull request merged**